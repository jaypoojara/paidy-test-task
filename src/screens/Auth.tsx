import React, { useEffect, useRef, useState } from "react";
import {
  Text,
  SafeAreaView,
  Linking,
  AppState,
  StyleSheet,
  Platform,
} from "react-native";
import * as LocalAuthentication from "expo-local-authentication";
import { StackActions, useNavigation } from "@react-navigation/native";
import Button from "../components/Button";
import { colors } from "../types/colors";
import { routesKey } from "../types/routesKey";

const Auth = () => {
  const appState = useRef(AppState.currentState);

  const navigation = useNavigation();

  const [enrolledLevel, setEnrolledLevel] = useState<number>(0);

  useEffect(() => {
    // Initialize the authentication settings
    init();

    // Add an AppState change listener to re-initialize on foregrounding
    const subscription = AppState.addEventListener("change", (nextAppState) => {
      if (
        appState.current.match(/inactive|background/) &&
        nextAppState === "active"
      ) {
        // Re-initialize when the app is brought to the foreground
        init();
      }
      appState.current = nextAppState;
    });

    // Clean up the listener when the component unmounts
    return () => {
      subscription.remove();
    };
  }, []);

  const init = async () => {
    // Get the enrolled authentication level
    const getEnrolledLevel = await LocalAuthentication.getEnrolledLevelAsync();

    // Update the state with the enrolled level
    setEnrolledLevel(getEnrolledLevel);
  };

  const onClickGoToSetting = () => {
    // Open the device's security settings
    if (Platform.OS === "ios") {
      Linking.openURL("App-Prefs://");
    } else {
      Linking.sendIntent("android.settings.SECURITY_SETTINGS");
    }
  };

  const onClickVerify = () => {
    // Perform authentication using biometrics
    const auth = LocalAuthentication.authenticateAsync({
      promptMessage: "Please authenticate",
      fallbackLabel: "Enter your device PIN to continue:",
    });
    auth.then((res) => {
      // Navigate to the Home screen if authentication succeeds
      if (res.success) {
        navigation.dispatch(StackActions.replace(routesKey.Home));
      }
    });
  };

  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.text}>Set Authentication to Proceed</Text>
      {enrolledLevel === 0 ? (
        // Display a button to navigate to security settings
        <Button onPress={onClickGoToSetting} title="Go to Settings" />
      ) : (
        // Display a button to verify authentication
        <Button onPress={onClickVerify} title="Verify" />
      )}
    </SafeAreaView>
  );
};

export default Auth;

// Styles for the screen
const styles = StyleSheet.create({
  container: {
    justifyContent: "flex-end",
    flex: 1,
    backgroundColor: colors.white,
    padding: 20,
    alignItems: "center",
  },
  text: {
    marginBottom: 15,
    textAlign: "center",
    fontSize: 18,
    fontWeight: "500",
  },
});
