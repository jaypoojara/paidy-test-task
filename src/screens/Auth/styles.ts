import { StyleSheet } from "react-native";
import { Colors } from "../../types/colors";

// Styles for the screen
const styles = StyleSheet.create({
  container: {
    justifyContent: "flex-end",
    flex: 1,
    backgroundColor: Colors.white,
    padding: 20,
    alignItems: "center",
  },
  text: {
    marginBottom: 15,
    textAlign: "center",
    fontSize: 18,
    fontWeight: "500",
  },
});

export { styles };
