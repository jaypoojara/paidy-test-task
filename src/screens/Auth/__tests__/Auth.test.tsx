import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import * as LocalAuthentication from "expo-local-authentication";
import Auth from "../index";

jest.mock("@react-navigation/native", () => ({
  useNavigation: jest.fn(() => ({
    navigate: jest.fn(),
  })),
}));

// Mock the LocalAuthentication module methods
jest.mock("expo-local-authentication", () => ({
  getEnrolledLevelAsync: jest.fn(),
  authenticateAsync: jest.fn(),
}));

describe("Auth Component", () => {
  it("renders correctly", () => {
    const element = render(<Auth />);
    expect(element).toMatchSnapshot();
  });

  it("renders correctly with no enrolled level", async () => {
    LocalAuthentication.getEnrolledLevelAsync.mockResolvedValue(0);

    const { getByText } = render(<Auth />);

    const textElement = getByText("Set Authentication to Proceed");
    expect(textElement).toBeDefined();

    const goToSettingsButton = getByText("Go to Settings");
    expect(goToSettingsButton).toBeDefined();

    fireEvent.press(goToSettingsButton);
  });
});
