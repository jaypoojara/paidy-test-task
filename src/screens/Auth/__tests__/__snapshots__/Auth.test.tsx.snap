// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Component renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "flex-end",
      "padding": 20,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 18,
        "fontWeight": "500",
        "marginBottom": 15,
        "textAlign": "center",
      }
    }
  >
    Set Authentication to Proceed
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#034096",
        "borderRadius": 100,
        "opacity": 1,
        "paddingHorizontal": 30,
        "paddingVertical": 12,
      }
    }
    testID="buttonTestID"
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Go to Settings
    </Text>
  </View>
</RCTSafeAreaView>
`;
